calour/io.py

```diff
@@ -316,102 +314,68 @@ def read(data_file, sample_metadata_file=None, feature_metadata_file=None,
+        sid, fid, data = _read_open_ms(data_file, rows_are_samples=False)
     elif data_file_type == 'openms_transpose':
-        sid, oid, data = _read_open_ms(data_file, rows_are_samples=True)
+        sid, fid, data = _read_open_ms(data_file, rows_are_samples=True)
     elif data_file_type == 'qiime2':
-        sid, oid, data, md = _read_qiime2(data_file)
+        sid, fid, data, fmd = _read_qiime2(data_file)
     elif data_file_type == 'tsv':
         df = pd.read_table(data_file, sep='\t', index_col=0)
         sid = df.columns.tolist()
-        oid = df.index.tolist()
+        fid = df.index.tolist()
         data = df.as_matrix().T
     else:
         raise ValueError('unkown data_file_type %s' % data_file_type)
 
-    sfilter = [True] * len(sid)
-    # load the sample metadata file
-    if sample_metadata_file is not None:
-        sample_metadata = _read_table(sample_metadata_file, encoding=encoding)
-        smid = set(sample_metadata.index)
-        sdid = set(sid)
-        diff = smid - sdid
-        if diff:
-            logger.warning('the samples are dropped because they have metadata but do not have data: %r' % diff)
-        sdiff = (sdid - smid)
-        if sdiff:
-            logger.warning('the samples have data but do not have metadata: %r' % sdiff)
-            sfilter = [i not in sdiff for i in sid]
-        # reorder the sample id to align with biom
-        sample_metadata = sample_metadata.loc[sid, ]
-        exp_metadata['map_md5'] = get_file_md5(sample_metadata_file, encoding=encoding)
-    else:
-        sample_metadata = pd.DataFrame(index=sid)
-        # duplicate the index to a column so we can select it
-        sample_metadata['id'] = sid
-
-    # load the feature metadata file
-    if feature_metadata_file is not None:
-        # reorder the feature id to align with that from biom table
-        feature_metadata = _read_table(feature_metadata_file, encoding=encoding)
-        fmid = set(feature_metadata.index)
-        fdid = set(oid)
-        diff = fmid - fdid
-        if diff:
-            logger.warning('the features are dropped because they have metadata but do not have data: %r' % diff)
-        fdiff = fdid - fmid
-        if fdiff:
-            logger.warning('the features have data but do not have metadata: %r' % fdiff)
-        feature_metadata = feature_metadata.loc[oid, ]
-    else:
-        feature_metadata = pd.DataFrame(index=oid)
-        feature_metadata['id'] = oid
-    if data_file_type == 'biom' and md is not None:
-        # combine it with the metadata
-        feature_metadata = pd.concat([feature_metadata, md], axis=1)
+    sample_metadata = _read_metadata(sid, sample_metadata_file, sample_metadata_kwargs)
+    feature_metadata = _read_metadata(fid, feature_metadata_file, feature_metadata_kwargs)
+
+    if fmd is not None:
+        # combine it with the feature metadata
+        feature_metadata = pd.concat([feature_metadata, fmd], axis=1)
```

Comment:
I like to have info about the difference between the mapping and biom table samples (i.e. how many were present in one but not the other). It lets you get some quick info about the success of sequencing and errors in the mapping file.
This part was deleted.... maybe we can add something for that end?

Reply:
this part is in the _read_metadata()